### ISSUE TITLE
[FEA] Support `COUNT_ALL` and `COUNT_VALID` as `reduce` aggregations

### ISSUE THREAD

**Issue by mythrocks** [rapidsai/cudf#13756]
This is a follow-up item that arose from https://github.com/rapidsai/cudf/pull/13727.

It turns out that `COUNT_ALL` and `COUNT_VALID` are not supported in `cudf::reduce()` as `reduce_aggregation`s.
(This is likely because their values can trivially computed from the results of `column::size()` and `column::null_count()`.)

However, this causes obfuscation in any code that attempts to dispatch reduce aggregations generically. E.g. [Here](https://github.com/rapidsai/cudf/pull/13727/files#diff-ee2fd30f5a96f800a146b5b8262244eeb73a7758834b6ff41f9b4343fe281c56R133).

It would be good to have `cudf::reduce()` support `COUNT_ALL` and `COUNT_VALID` natively, so as not to require special handling in the calling code.

**Comment by davidwendt** [comment]
I disagree with this change. I think [`cudf::reduce`](https://docs.rapids.ai/api/libcudf/stable/group__aggregation__reduction.html#ga48cc284a9371891975a6219cd3d2e36e) has become too monolithic and difficult to maintain as more aggregations are added. Special rules have to be applied and adjusted (and documented) concerning combinations of parameters and parameter types. For example, certain aggregations do not accept `init` values and others only accept specific `output_type`s.

I'd like to see libcudf split up the reduce algorithms into functional names like `cudf::reduction::any(), cudf::reduction::all(), cudf::reduction::max(), etc`
Internally the reduce/aggregation types are implemented as `cudf::reduction::detail::any(), cudf::reduction::detail::all(), cudf::reduction::detail::max(), etc`
This will make it easier to maintain and document as well as add new reductions in the future without breaking the monolithic API. I can open a separate issue to further discuss this more practically. 

Regardless, I don't think we should provide a convenience wrapper on the existing `cudf::column::size()` or `cudf::column::null_count()`

**Comment by ttnghia** [comment]
From my perspective, putting the commonly used feature at the lower level (libcudf) is better than at the application level. Why? Because we can just call libcudf API and avoid reimplementing that same feature multiple times in multiple applications.

**Comment by bdice** [comment]
If we were able to eliminate the enum `aggregation::Kind` by moving to a functional interface like `cudf::reduction::any()` then I would support the expectation that users should call `cudf::column::size()` as @davidwendt proposes. I do think that's the direction we ought to go, but I haven't scoped any of what that work would require.

However, as it stands with `aggregation::Kind` today, I am not a fan of a design that allows only certain aggregations to be used as a reduction. If the reduction API accepts an enum `aggregation::Kind`, it should support all possible values of that enum even if they are redundant with an existing function, for "application level" / "special casing" reasons as @ttnghia and @mythrocks pointed out above.

Given the tension between these two potential next steps, I would encourage spending more effort towards the functional API that @davidwendt proposed, but I wouldn't oppose expanding the existing reduction API's `aggregation::Kind` support if we think there is enough justification, particularly potential for reuse in other parts of libcudf's aggregation code paths.
